pinn/networks/pinet2_modularized.py

```diff
@@ -0,0 +1,313 @@
+            pp1_nodes[-1] = ii_nodes[-1] * self.n_layers
+            self.invar_p1_layer = InvarLayer(pp_nodes, pi_nodes, ii1_nodes, **kwargs)
+            self.pp_layer = FFLayer(pp1_nodes, **kwargs)
+
+        if rank >= 3:
+            self.equivar_p3_layer = EquiVarLayer(ppx_nodes, weighted=weighted, **kwargs)
+
+        if rank >= 5:
+            self.equivar_p5_layer = EquiVarLayer(ppx_nodes, weighted=weighted, **kwargs)
+
+        self.scale3_layer = ScaleLayer()
+        self.scale5_layer = ScaleLayer()
+
+    def call(self, tensors):
+
+        ind_2, basis, px, diffs = tensors
+
+        p1, i1 = self.invar_p1_layer([ind_2, px[0], basis])
+
+        i1s = tf.split(i1, self.n_layers, axis=-1)
+        px_list = [p1]
+        ix_list = [i1]
+
+        if self.rank >= 3:
+            p3, i3, dotted_p3 = self.equivar_p3_layer(
+                [ind_2, px[1], i1s[1], diffs[0]]
+            )  # NOTE: use same i1 branch for diff_px and px, same result as separated i1
+            px_list.append(dotted_p3)
+            ix_list.append(i3)
+
+        if self.rank >= 5:
+            p5, i5, dotted_p5 = self.equivar_p5_layer([ind_2, px[2], i1s[2], diffs[1]])
+            px_list.append(dotted_p5)
+            ix_list.append(i5)
+
+        p1t1 = self.pp_layer(
+            tf.concat(
+                px_list,
+                axis=-1,
+            )
+        )
+
+        pxt1 = tf.split(p1t1, self.n_layers, axis=-1)
+        pxt1_list = [pxt1[0]]
+
+        if self.rank >= 3:
+            p3t1 = self.scale3_layer([p3, pxt1[1]])
+            pxt1_list.append(p3t1)
+
+        if self.rank >= 5:
+            p5t1 = self.scale5_layer([p5, pxt1[2]])
+            pxt1_list.append(p5t1)
+
+        return pxt1_list, ix_list
+
+
+class PreprocessLayer(tf.keras.layers.Layer):
+    def __init__(self, rank, atom_types, rc):
+        super(PreprocessLayer, self).__init__()
+        self.rank = rank
+        self.embed = AtomicOnehot(atom_types)
+        self.nl_layer = CellListNL(rc)
+
+    def call(self, tensors):
+        tensors = tensors.copy()
+        for k in ["elems", "dist"]:
+            if k in tensors.keys():
+                tensors[k] = tf.reshape(tensors[k], tf.shape(tensors[k])[:1])
+        if "ind_2" not in tensors:
+            tensors.update(self.nl_layer(tensors))
+            tensors["p1"] = tf.cast(  # difference with pinet: prop->p1
+                self.embed(tensors["elems"]), tensors["coord"].dtype
+            )
+
+        return tensors
+
+
+class PiNet2(tf.keras.Model):
+    """This class implements the Keras Model for the PiNet network."""
+
+    def __init__(
+        self,
+        atom_types=[1, 6, 7, 8],
+        rc=4.0,
+        cutoff_type="f1",
+        basis_type="polynomial",
+        n_basis=4,
+        gamma=3.0,
+        center=None,
+        pp_nodes=[16, 16],
+        pi_nodes=[16, 16],
+        ii_nodes=[16, 16],
+        out_nodes=[16, 16],
+        out_units=1,
+        out_extra={},
```

Comment:
"out_extra" and how to use it is not in the documentation of this function. The same goes for "out_units".